user_profile_modal: Abbreviate long URLs of custom fields.

Ddharmani3 opened the pull request:
<!-- Describe your pull request here.-->
Abbreviated long URL for custom field in the user profile.
Added copy to clipboard icon at end of the field with tippy on hover that says `Copy URL`.
Clicking on the copied button shows a `Copied!` tippy tooltip for 1 sec.

Fixes: <!-- Issue link, or clear description.-->
#21680

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

https://user-images.githubusercontent.com/64723994/228928380-e2f95be3-110e-42a5-be19-f47606efe509.mp4

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

Ddharmani3 commented:
@alya, I have implemented all 3 points you mentioned here. But I want to share 4 tooltip ideas for showing long_url:(screenshots below of URL with maximum characters)
1) 700px max-width tooltip:
![image](https://user-images.githubusercontent.com/64723994/228930922-ff1376fd-1981-415b-bc60-d49d03f8e770.png)
on mobile size, it gets out of the screen.
2) 400pxmax-width tooltip with 7 lines verticle clamping:
![image](https://user-images.githubusercontent.com/64723994/228931458-84f0856f-839a-4161-b804-bff1688e849e.png)
on mobile size, it is fully visible.
3) bootstrap tooltip with title attribute: unable to capture ss but looks very big.
4) I think this approach is most suitable as the main domain name is visible in the visible part of the URL, and if the user want to see more or use it there is a copy button anyways. so I don't think showing the full URL on hover will have any use especially when the tooltip is getting so big and making the UX worse.

what is your thought about which tooltip to use?

alya commented:
Sure, not using a tooltip seems fine to me.

alya commented:
In manual testing, I get this error when clicking on the "Copy URL" button:

<img width="967" alt="Screen Shot 2023-03-30 at 11 52 18 PM" src="https://user-images.githubusercontent.com/2090066/229045533-dba5db1f-3e97-4d30-8d22-c6b519221c65.png">

alya commented:
We should also use a Tippy tooltip for the "Copy URL" button:

<img width="131" alt="Screen Shot 2023-03-30 at 11 52 59 PM" src="https://user-images.githubusercontent.com/2090066/229045640-6b6599db-e661-47fd-9821-52cb4288b7e2.png">

alya commented:
For a shorter URL, the "copy" button should be next to the URL, not off on the right:

<img width="528" alt="Screen Shot 2023-03-30 at 11 54 40 PM" src="https://user-images.githubusercontent.com/2090066/229045802-6b2d6a0d-c23a-4390-adbc-33b494c25333.png">

alya commented:
That's all I've got at this point.

Ddharmani3 commented:
@alya, I have addressed all the feedback:
I have removed the tooltip on the URL,
added a tippy tooltip on hover for the Copied button,
and fixed the alignment of the copied button to be at the end of the URL, not off on the right. 
These changes are now ready for review.

I am not sure about the error you described here; it's not showing me any error on manual testing locally. 

https://user-images.githubusercontent.com/64723994/229057827-85609f71-f8d2-4389-be60-ee1bac27e70f.mp4

Ddharmani3 commented:
@alya This PR is ready for review now.

amanagr commented on web/templates/user_custom_profile_fields.hbs:
```diff
@@ -4,26 +4,31 @@
-                {{/if}}
-                {{this.value}}
-            </a>
+        <a tabindex="0" href="{{this.link}}" target="_blank" rel="noopener noreferrer" class="value custom_profile_fields_link {{#if ../for_user_info_popover}}tippy-zulip-tooltip{{/if}}" data-tippy-content="{{this.name}}">
+            {{#if (eq this.subtype "github") }}
+            <i class="fa fa-github" aria-hidden="true"></i>
+            {{else if (eq this.subtype "twitter") }}
+            <i class="fa fa-twitter" aria-hidden="true"></i>
+            {{/if}}
+            {{this.value}}
+        </a>
         {{else if this.is_user_field}}
-            <div class="pill-container not-editable" data-field-id="{{this.id}}">
-                <div class="input" contenteditable="false" style="display: none;"></div>
-            </div>
+        <div class="pill-container not-editable" data-field-id="{{this.id}}">
+            <div class="input" contenteditable="false" style="display: none;"></div>
+        </div>
         {{else}}
-            {{#if this.rendered_value}}
-            <span class="value rendered_markdown {{#if ../for_user_info_popover}}tippy-zulip-tooltip{{/if}}" data-tippy-content="{{this.name}}">{{rendered_markdown this.rendered_value}}</span>
-            {{else}}
-            <span class="value {{#if ../for_user_info_popover}}tippy-zulip-tooltip"{{/if}} data-tippy-content="{{this.name}}">{{this.value}}</span>
-            {{/if}}
+        {{#if this.rendered_value}}
+        <span class="value rendered_markdown {{#if ../for_user_info_popover}}tippy-zulip-tooltip{{/if}}" data-tippy-content="{{this.name}}">{{rendered_markdown this.rendered_value}}</span>
+        {{else}}
+        <span class="value {{#if ../for_user_info_popover}}tippy-zulip-tooltip"{{/if}} data-tippy-content="{{this.name}}">{{this.value}}</span>
+        {{/if}}
```
Can we remove the unrelated indentation changes from here?

amanagr commented:
@Ddharmani3 LGTM! Left a minor comment.

amanagr commented on web/src/user_profile.js:
```diff
@@ -224,6 +226,25 @@ export function show_user_profile(user, default_tab_key = "profile-tab") {
     $(".tabcontent").hide();
 
+    const copy_url_tooltip = tippy("#copy_custom_field_url i", {
+        content: "Copied!",
```
Oh! This needs to be marked for translation.

amanagr commented on web/templates/user_custom_profile_fields.hbs:
```diff
@@ -4,7 +4,12 @@
+            <div class="custom_user_URL_field">
+                <a tabindex="0" href="{{this.value}}" target="_blank" rel="noopener noreferrer" id="custom_profile_fields_long_url" class="value custom_profile_fields_link {{#if ../for_user_info_popover}}tippy-zulip-tooltip{{/if}}">{{this.value}}</a>
+                <button type="submit" id="copy_custom_field_url" class="tippy-zulip-tooltip" data-tippy-content="{{t 'Copy URL' }}">
+                    <i class="fa fa-clipboard copy-blue"></i>
```
I think we should be using `{{> copy_to_clipboard_svg }}` here instead. Also, another commit to replace it in `bot_avatar_row.hbs` would be nice so that we use the same copy icon across the app.

amanagr reviewed (commented):
@Ddharmani3 left some comments. Looks great!

Ddharmani3 commented:
@amanagr updated the PR to resolve the conflicts and also addressed the feedback. This PR is ready for review now.

amanagr commented:
LGTM

amanagr commented:
LGTM

Ddharmani3 commented:
Updated the PR to fix some Issues, as I noticed that this same field was being used in the user card popover too and this change was affecting user card popover. So, fixed that and now added the same copy icon there too.

Here are the updated screenshots/screen captures:

<details>
  <summary>User card popover copy icon(2nd commit)</summary>
  
![image](https://github.com/zulip/zulip/assets/64723994/484336ba-3f9f-4558-83f8-c3b091f98e6d)

Uploading (1) Recent conversations - Zulip Dev - Zulip - Google Chrome 2023-09-18 16-18-29.mp4…
 
</details>

<details>
  <summary>User profile modal (2nd commit)</summary>
  
![image](https://github.com/zulip/zulip/assets/64723994/0fdcccf8-116c-41c4-a87b-88440a0e5ae2)


https://github.com/zulip/zulip/assets/64723994/ae064be8-095d-4b00-9535-5e5549dabfd5

</details>

<details>
  <summary>Updated copy Icon in Bot settings (3rd commit)</summary>

  Before:
![image](https://github.com/zulip/zulip/assets/64723994/fc325f26-525a-4afa-8bd7-d34a8dc32319)

After:
 
![image](https://github.com/zulip/zulip/assets/64723994/59ebd219-62dc-4e0a-9d0c-7987a5cbf8cf)

</details>

amanagr commented:
Oh! Thanks for fixing. LGTM

timabbott commented:
Merged, thanks @Ddharmani3 and @alya!
